src/us/kbase/workspace/database/provenance/Provenance.java

```diff
@@ -0,0 +1,183 @@
+package us.kbase.workspace.database.provenance;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Instant;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+import us.kbase.workspace.database.WorkspaceUser;
+
+/** Provenance for a data object that describes how that data object was created. */
+public class Provenance {
+	
+	private final List<ProvenanceAction> actions;
+	private final WorkspaceUser user;
+	private final Instant date;
+	private final long wsid;
+	
+	private Provenance(
+			final WorkspaceUser user,
+			final Instant date,
+			final long wsid,
+			final List<ProvenanceAction> actions) {
+		this.actions = actions;
+		this.user = user;
+		this.date = date;
+		this.wsid = wsid;
+	}
+
+	/** Get the user that created the data.
+	 * @return the user.
+	 */
+	public WorkspaceUser getUser() {
+		return user;
+	}
+
+	/** Get the time the data was saved to the workspace.
+	 * @return the time.
+	 */
+	public Instant getDate() {
+		return date;
+	}
+
+	/** Get the original workspace in which the data was saved.
+	 * Data saved before version 0.4.1 of the workspace never have this field.
+	 * @return the original workspace ID, if present.
+	 */
+	public Optional<Long> getWorkspaceID() {
+		return wsid < 1 ? Optional.empty() : Optional.ofNullable(wsid);
+	}
+
+	/** Get the actions that took place while creating the data object.
+	 * @return the actions.
+	 */
+	public List<ProvenanceAction> getActions() {
+		return Common.getList(actions);
+	}
+
+	/** Update the workspace ID for the provenance. This returns this provenance instance if
+	 * the workspace ID doesn't change, or a new provenance instance if the incoming workspace
+	 * ID is different.
+	 * @param workspaceID the new workspace ID.
+	 * @return this or a new Provenance.
+	 */
+	public Provenance updateWorkspaceID(final Long workspaceID) {
+		if (workspaceID == null) {
+			if (wsid == -1) {
+				return this;
+			}
+			return new Provenance(user, date, -1, actions);
+		}
+		if (workspaceID < 1) {
+			throw new IllegalArgumentException("workspace ID must be > 0");
+		}
+		if (wsid == workspaceID) {
+			return this;
+		}
+		return new Provenance(user, date, workspaceID, actions);
+	}
+
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ((actions == null) ? 0 : actions.hashCode());
+		result = prime * result + ((date == null) ? 0 : date.hashCode());
+		result = prime * result + ((user == null) ? 0 : user.hashCode());
+		result = prime * result + (int) (wsid ^ (wsid >>> 32));
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
```

Comment:
Glad to read in the comments of the previous PR that you don't have to write this all by hand. That would be very tedious indeed.